pageserver/src/layered_repository/inmemory_layer.rs

```diff
@@ -111,6 +117,50 @@ impl InMemoryLayerInner {
             panic!("could not find seg size in in-memory layer");
         }
     }
+
+    ///
+    /// Read a page version from the ephemeral file.
+    ///
+    fn read_pv(&self, off: u64) -> Result<PageVersion> {
+        let mut buf = Vec::new();
+        self.read_pv_bytes(off, &mut buf)?;
+        Ok(PageVersion::des(&buf)?)
+    }
+
+    ///
+    /// Read a page version from the ephemeral file, as raw bytes, at
+    /// the given offset.  The bytes are read into 'buf', which is
+    /// expanded if necessary. Returns the size of the page version.
+    ///
+    fn read_pv_bytes(&self, off: u64, buf: &mut Vec<u8>) -> Result<usize> {
+        // read length
+        let mut lenbuf = [0u8; 4];
+        self.file.read_exact_at(&mut lenbuf, off)?;
+        let len = u32::from_ne_bytes(lenbuf) as usize;
```

Comment:
There's an invariant here that the VecMap values always point to a size. If we accidentally break it we might end up corrupting data without noticing. A few solutions that come to mind:
a) Use some type trickery to prevent bugs, with `struct HeaderOffset(u64);` or something similar.
b) Add abstraction for MemoryBackedMap<KeyType> that implements only read and write. This way we effectively divide the problem in half: Imlement correct file ops without worrying about pages and postgres details; Implement InMemoryLater without worrying about corruption.

Not necessary to address now, just something to think about.